docs/src/tutorial-nondeterministic-variables.md

```diff
@@ -1,95 +1,114 @@
 
-Kani also includes an **unsafe** method to generate unconstrained nondeterministic variables which do not take type invariants into consideration.
-As with any unsafe method in Rust, users have to make sure the right guardrails are put in place to avoid undesirable behavior.
+## Bounding nondeterministic variables
 
-That said, there may be cases where you want to verify your code taking into consideration that some inputs may contain invalid data.
+There is no implementation of `kani::any()` for more complex data structures like `Vec`, however.
+The trouble with a nondeterministic vector is that you usually need to _bound_ the size of the vector, for the reasons we investigated in the [last chapter](./tutorial-loop-unwinding.md).
+The `kani::any()` function does not have any arguments, and so cannot be given an upper bound.
 
-Let's see what happens if we modify our verification harness to use the unsafe method `kani::any_raw()` to generate the updated value.
+You can use `kani::any()` for `[T; N]` (if implemented for `T`).
+But this the type of an exactly-sized array.
+Again, `kani::any()` does not have an implementation for an unsized slice (like `[T]`).
 
-```rust
-{{#include tutorial/arbitrary-variables/src/inventory.rs:unsafe_update}}
-```
+This does not mean you cannot have a nondeterministic vector.
+It just means you have to construct one.
+Our example proof harness above constructs a nondeterministic `Inventory` of size `1`, simply by starting with the empty `Inventory` and inserting a nondeterministic entry.
 
-We commented out the assertion that the underlying `u32` variable cannot be `0`, since this no longer holds.
-The verification will now fail showing that the `inventory.get(&id).unwrap()` method call can panic.
+## Custom nondeterministic types
 
-This is an interesting issue that emerges from how `rustc` optimizes the memory layout of `Option<NonZeroU32>`.
-The compiler is able to represent `Option<NonZeroU32>` using `32` bits by using the value `0` to represent `None`.
+There's nothing special about `kani::any()` except that Kani ships with implementations for a few types where we can guarantee safety.
+When you need nondeterministic variables of types that `kani::any()` cannot construct, you have two options:
 
-You can try it out by running the example under [`arbitrary-variables`](https://github.com/model-checking/kani/tree/main/docs/src/tutorial/arbitrary-variables/):
+1. Implement `kani::Arbitrary` for your type, so you can use `kani::any()`.
+2. Just write a function.
 
-```
-cargo kani --harness unsafe_update
-```
+The advantage of the first approach is that it's simple and conventional.
+It also means that in addition to being able to use `kani::any()` with your type, you can also use it with `Option<MyType>` (for example).
 
-## Safe nondeterministic variables for custom types
+The advantage of the second approach is that you're able to pass in parameters, like bounds on the size of the data structure.
```

Comment:
Here you mention the parameter, so you could just replace the previous section by giving a bit more details on the second approach. Maybe add an example.